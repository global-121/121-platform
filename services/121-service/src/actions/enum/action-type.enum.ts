export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
