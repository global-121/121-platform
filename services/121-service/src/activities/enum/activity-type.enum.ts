export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
