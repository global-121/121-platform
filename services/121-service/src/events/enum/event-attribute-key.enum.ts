export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
