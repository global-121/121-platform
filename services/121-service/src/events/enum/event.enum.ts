export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
