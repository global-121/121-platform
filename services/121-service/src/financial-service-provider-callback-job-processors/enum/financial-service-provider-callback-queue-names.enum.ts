export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
