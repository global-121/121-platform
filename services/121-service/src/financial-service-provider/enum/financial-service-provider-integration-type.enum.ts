export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
