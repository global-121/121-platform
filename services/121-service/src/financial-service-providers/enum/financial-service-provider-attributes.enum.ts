export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
