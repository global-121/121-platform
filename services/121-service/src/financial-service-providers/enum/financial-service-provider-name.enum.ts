export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
