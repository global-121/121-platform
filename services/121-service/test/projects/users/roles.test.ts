import { HttpStatus } from '@nestjs/common';

import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import { DefaultUserRole } from '@121-service/src/user/user-role.enum';
import {
  getAccessToken,
  getServer,
  resetDB,
} from '@121-service/test/helpers/utility.helper';

describe('Projects / Users / Roles', () => {
  const projectId = 1;
  const userId = 2;
  const fixtureUserRoles = [
    {
      id: 2,
      role: DefaultUserRole.ProjectAdmin,
      label: 'Project Admin',
    },
    {
      id: 3,
      role: DefaultUserRole.View,
      label: 'Only view data, including Personally Identifiable Information',
    },
    {
      id: 5,
      role: DefaultUserRole.CvaManager,
      label: 'Cash Assistance Project Manager',
    },
  ];

  let accessToken: string;

  beforeEach(async () => {
    await resetDB(SeedScript.testMultiple, __filename);
    accessToken = await getAccessToken();
  });

  it('should only return user roles and assignments to specific project assignments', async () => {
    // Arrange
    const testUserRoles = fixtureUserRoles;

    // Act
    const response = await getServer()
      .get(`/projects/${projectId}/users/${userId}`)
      .set('Cookie', [accessToken]);

    // Assert
    expect(response.status).toBe(HttpStatus.OK);
    expect(response.body.roles.length).toBe(1);
    expect(response.body.roles[0].role).toBe(testUserRoles[0].role);
    expect(response.body.roles[0].id).toBe(testUserRoles[0].id);
    expect(response.body.roles[0].label).toBe(testUserRoles[0].label);
  });

  it('should return user roles after update to specific project assignments', async () => {
    // Arrange
    const testUserRoles = fixtureUserRoles;
    const testRoles = {
      roles: ['project-admin', 'cva-manager'],
      scope: 'test',
    };

    // Act
    const response = await getServer()
      .put(`/projects/${projectId}/users/${userId}`)
      .set('Cookie', [accessToken])
      .send(testRoles);

    // Assert
    expect(response.status).toBe(HttpStatus.OK);
    expect(response.body.roles.length).toBe(2);
    expect(response.body.roles[0].role).toBe(testUserRoles[0].role);
    expect(response.body.roles[0].id).toBe(testUserRoles[0].id);
    expect(response.body.roles[0].label).toBe(testUserRoles[0].label);
  });

  it('should return an error when a user tries to assign roles to themselves', async () => {
    // Arrange
    const selfUserId = 1; // Assuming the logged-in user has ID 1
    const testRoles = {
      roles: ['project-admin'],
      scope: 'test',
    };

    // Act
    const response = await getServer()
      .put(`/projects/${projectId}/users/${selfUserId}`)
      .set('Cookie', [accessToken])
      .send(testRoles);

    // Assert
    expect(response.status).toBe(HttpStatus.FORBIDDEN);
    expect(response.body.message).toMatchSnapshot();
  });

  it('should return user roles after add new role to specific project assignment', async () => {
    // Arrange
    const testUserRoles = fixtureUserRoles;
    const testRoles = {
      rolesToAdd: ['view'],
      scope: 'test',
    };

    // Act
    const response = await getServer()
      .patch(`/projects/${projectId}/users/${userId}`)
      .set('Cookie', [accessToken])
      .send(testRoles);

    // Assert
    expect(response.status).toBe(HttpStatus.OK);
    expect(response.body.roles.length).toBe(2);
    expect(response.body.roles[1].role).toBe(testUserRoles[1].role);
    expect(response.body.roles[1].id).toBe(testUserRoles[1].id);
    expect(response.body.roles[1].label).toBe(testUserRoles[1].label);
  });

  it('should return user roles after delete roles from specific project assignment', async () => {
    // Arrange
    const testUserRoles = fixtureUserRoles;
    const testRoles = {
      rolesToDelete: ['view', 'cva-manager'],
    };

    // Act
    const response = await getServer()
      .delete(`/projects/${projectId}/users/${userId}`)
      .set('Cookie', [accessToken])
      .send(testRoles);

    // Assert
    expect(response.status).toBe(HttpStatus.OK);
    expect(response.body.roles.length).toBe(1);
    expect(response.body.roles[0].role).toBe(testUserRoles[0].role);
    expect(response.body.roles[0].id).toBe(testUserRoles[0].id);
    expect(response.body.roles[0].label).toBe(testUserRoles[0].label);
  });
});
