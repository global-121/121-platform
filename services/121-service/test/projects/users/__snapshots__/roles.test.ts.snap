// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects / Users / Roles should return an error when a user tries to assign roles to themselves 1`] = `"You cannot update your own project assignment"`;
